# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit/MacroAssembler.cpp
# Commit: 3cfa16c9f9a9
# Full Hash: 3cfa16c9f9a97c5963b31a9ca080df29e772836c
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2020-10-30 16:09:44
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Part 3: Implement the algorithm for obtaining tls and use it for wasm signal handling. r=lth
#   
#   
#   This is the third part of series of patches to Frame without tls pointer.
#   Here we preserve initial tls in all entry stubs and then use it to find a proper tls instance for a given frame.
# ==============================================================================

diff -r 50772b4ff6aa -r 3cfa16c9f9a9 js/src/jit/MacroAssembler.cpp
--- a/js/src/jit/MacroAssembler.cpp	Fri Oct 23 18:17:37 2020 +0000
+++ b/js/src/jit/MacroAssembler.cpp	Fri Oct 23 18:17:37 2020 +0000
@@ -3796,6 +3796,7 @@
            Address(getStackPointer(), WasmCallerTLSOffsetBeforeCall));
   storePtr(WasmTlsReg,
            Address(getStackPointer(), WasmCalleeTLSOffsetBeforeCall));
+
   if (instanceArg.kind() == ABIArg::GPR) {
     loadPtr(Address(WasmTlsReg, offsetof(wasm::TlsData, instance)),
             instanceArg.gpr());
@@ -4656,6 +4657,18 @@
 template void AutoGenericRegisterScope<Register>::reacquire();
 template void AutoGenericRegisterScope<FloatRegister>::reacquire();
 
+wasm::TlsData* ExtractCallerTlsFromFrameWithTls(wasm::Frame* fp) {
+  return *reinterpret_cast<wasm::TlsData**>(
+      reinterpret_cast<uint8_t*>(fp) + sizeof(wasm::Frame) + ShadowStackSpace +
+      wasm::FrameWithTls::callerTLSOffset());
+}
+
+wasm::TlsData* ExtractCalleeTlsFromFrameWithTls(wasm::Frame* fp) {
+  return *reinterpret_cast<wasm::TlsData**>(
+      reinterpret_cast<uint8_t*>(fp) + sizeof(wasm::Frame) + ShadowStackSpace +
+      wasm::FrameWithTls::calleeTLSOffset());
+}
+
 #endif  // DEBUG
 
 }  // namespace jit